### ISSUE TITLE
Shorter syntax for unwrapping and casting through any

### ISSUE THREAD

**Issue by halaei** [facebook/flow#8786]
## Proposal

The rules of type checking in Flow forces us to write safer code. But when working with union types, sometimes it leads to code that is longer than it can be, specially when it comes to casting through any. I hope if we can have a shorter, more readable casting syntax. I propose to use 2 new operators, `!:` and `!.`.

### Use case

```js
type Track = {|
    ...,
    liked: boolean,
|};
type State = {|
    track: void|Track,
|};
const state : State = {
    track: undefined,
};
```
Current syntax for casting through any:
```js
((state.track: any): Track).liked = ! ((state.track: any): Track).liked;
```
A shorter syntax for casting through any, which is more readable and have the same semantics:
```js
(state.track !: Track) = ! (state.track !: Track);
```
Alternatively, we could tell Flow that we know what we do and we are sure at this point that `state.track` is of a type that matches our expectation, having a liked property. Flow can verify that it is a possibility, so it stop giving us errors:
```js
state.track!.liked = !state.track!.liked;
```

I like the readability of using the proposed `!.` operator, but it can be harder to implement or agree on exact specifications. On the other hand, `!:` can be implemented a lot easier because it doesn't actually introduce a new concept. It's just a short syntax.

**Comment by gnprice** [comment]
> working with union types […] casting through any.

This is not a recommended way to use Flow. You will get more benefit from Flow if you adapt your code into patterns that work with the type-checker rather than against it. In particular, this means not using `any`.

One useful way to manage that sort of migration across your codebase is with "Flow Strict": https://flow.org/en/docs/strict/

In general the direction of Flow is toward less use of `any`, not more. If you haven't already read it, this blog post is an important one to read for anyone responsible for the use of Flow in a codebase that isn't at Facebook:
  https://medium.com/flow-type/clarity-on-flows-direction-and-open-source-engagement-e721a4eb4d8b
In particular:
> Today, almost every JS file in Facebook’s codebase has Flow enabled, and over 90% are at the higher Flow Strict standard.

That makes it very unlikely that the Flow team will be interested in features for more-convenient use of `any`.

---

About what to do instead of `any`, in your example:
```js
type State = {|
    track: void|Track,
|};
const state : State = // …

((state.track: any): Track).liked = ! ((state.track: any): Track).liked;
```

what's happening in that last line is that it's making an assumption that `state.track` is not in fact void, but rather is of type `Track`. Instead of using `any`, a good pattern would be to revise the code so that it directly expresses why that assumption is true.

For example, perhaps you have control flow that checks that it's not void. In that case Flow can track the effect of that check, as a [type refinement](https://flow.org/en/docs/lang/refinements/). So your code might look like:
```js
  const { track } = state;
  if (!track) {
    return;
  }
  track.liked = ! track.liked;
```
The type-checker will accept that as correct.

Alternatively, perhaps you know it's not void based on some broader invariant in your program. You can express that with an assertion:
```js
  if (!track) {
    throw new Error('track missing');
  }
  track.liked = ! track.liked;
```

Compared to the `any` approach, this has the important advantage that if for some reason the assumption is ever wrong, you'll get a specific exception telling you so, instead of having your program go wrong in some subtle way. Its other important advantage is that it doesn't suppress *other* unrelated errors -- for example, perhaps you make a new "track" type and `state.track` is now `void | PartialTrack`. With the `any` cast, the type-checker can't point out that you need to update this site.

**Comment by halaei** [comment]
Thanks @gnprice I get your point. I guess my confusion comes with the fact that sometimes flow isn't smart enough to detect types, so I went for casting through any instead of adding redundant validations. To me it sounds simpler and faster, at least maybe in some cases. Examples are:

1. A container can't be undefined when it's content is already defined. But Flow doesn't get it. It is a Flow bug in my opinion:
```js
const tracks: void|Track[] = undefined;
function like(index) {
    const track = tracks?.[index];
    if (track) {
        // Here tracks can't be undefined but Flow doesn't get it.
        // Error: Cannot assign object literal to `tracks[index]` because  undefined [1] does not have properties.
        tracks[index] = {...track, liked: true};
    }
}
```
2. Having validation in a caller function
```js
function handleToggleLike() {
    if (state.track) toggleLike();
}
function toggleLike() {
     // Is there any way that Flow can get at this point state.track can't be undefined without writing extra code?
     // If not, can we have shorter syntax for casting through any? :)
     state.track = ! state.track;
}
```
3. Our `toggleLike()` function is only triggered by a UI event on a component that only exists when `state.track` is set. But how we can let Flow know that?

**Comment by gkz** [comment]
I appreciate the desire for improved refinement tracking logic. We aren't going to make doing unsafe casts easier as a workaround however, as it goes against our direction of increased type safety. We encourage you to write the extra code necessary instead.

**Comment by halaei** [comment]
Thanks @gkz 
I hope for improvement in refinement tracking logic as you named it. I am getting the concepts better now and was successful in removing any in an example application. I just want to point out that I came across another issue in my example that seems to be an old reported issue #3073 regarding array destructing. I added some extra checks to see if they work but they didn't:
```js
const tracks: Track[] = [];

if (tracks[0] && tracks[1] && tracks.length > 1) {
     // Error Unsupported expression pattern in destructuring:
    [tracks[0], tracks[1]] = [tracks[1], tracks[0]];
}
```
Am I supposed to just add `// $FlowFixMe` for now?
